### PR TITLE
Add BoundedInts to DafnyStdLibs

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Description
1. Add `BoundedInts.dfy`, `BoundedInts.md` and `BoundedIntExamples.dfy` from https://github.com/dafny-lang/libraries to https://github.com/dafny-lang/dafny/tree/master/Source/DafnyStandardLibraries
1. For Go, allows customizing the dummy value that is used for an `{:extern}` module, using a new undocumented, only internally used, `{:member <name>, <isType>}` attribute. This is necessary to let the `BoundedIntExamples.dfy` pass, because `FunctionExamples.dfy`, which is part of the same project, does not use the `{:extern}` module, so Go gives an unused import error. In the future I expect that we change our Go compiler so it only emits imports that are used, which makes the `{:member}` attribute obsolete and allows us to remove support for it from our codebase without breaking anyone.

### How has this been tested?
- The examples file serves as a test. It contains an extern method that demonstrates which native types the Dafny compiler picks when using `int32` in Dafny.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyStandardLibraries/examples/WrappersExamples.dfy [hunk 1]
```diff
@@ -1,172 +1,168 @@
-module Demo {
+import opened DafnyStdLibs.Wrappers
```
Comment (robin-aws) [inline Source/DafnyStandardLibraries/examples/WrappersExamples.dfy:1]: Good change, but wow did GitHub ever mess up the diff something terrible. :)

File: Source/DafnyStandardLibraries/examples/BoundedIntExamples.dfy [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+method {:test} UseExterns() {
+  var squareOf8 := Externs.NonDefault.SquareNativeInt(8);
+  expect squareOf8 == 64;
+}
+
+module {:extern} Externs {
+  import opened DafnyStdLibs.BoundedInts
+  class {:extern} NonDefault {
```
Comment (robin-aws) [inline Source/DafnyStandardLibraries/examples/BoundedIntExamples.dfy:19]: Why "NonDefault"?

File: Source/DafnyStandardLibraries/Makefile [hunk 3]
```diff
@@ -23,11 +23,11 @@ update-binary: build-binary
...
-	$(DAFNY) test -t:cs build/stdlibexamples.doo --output:build/stdlibexamples
-	$(DAFNY) test -t:java build/stdlibexamples.doo --output:build/stdlibexamples
-	$(DAFNY) test -t:go build/stdlibexamples.doo --output:build/stdlibexamples
-	$(DAFNY) test -t:py build/stdlibexamples.doo --output:build/stdlibexamples
-	$(DAFNY) test -t:js build/stdlibexamples.doo --output:build/stdlibexamples
+	$(DAFNY) test -t:cs build/stdlibexamples.doo examples/BoundedInts/NonDefault.cs --output:build/stdlibexamples
+	$(DAFNY) test -t:java build/stdlibexamples.doo examples/BoundedInts/Externs/NonDefault.java --output:build/stdlibexamples
+	$(DAFNY) test -t:go build/stdlibexamples.doo examples/BoundedInts/NonDefault.go --output:build/stdlibexamples
+	$(DAFNY) test -t:py build/stdlibexamples.doo examples/BoundedInts/NonDefault.py --output:build/stdlibexamples
+	$(DAFNY) test -t:js build/stdlibexamples.doo examples/BoundedInts/NonDefault.js --output:build/stdlibexamples
```
Comment (robin-aws) [inline Source/DafnyStandardLibraries/Makefile:30]: Fine for this change but probably worth replacing with a `find` of all files of the right extension in the next one, and before too long with project file support for this. :)

**Review by robin-aws — COMMENTED** [review]
Looks great and I like the testing approach! Just a couple of thoughts for later and a question.

**Comment by robin-aws** [comment]
Almost forgot, make sure you update https://github.com/dafny-lang/dafny/tree/master/Source/DafnyStandardLibraries#available-libraries too!

Might be worth adding a tiny `Functions.md` for that too.

File: Source/DafnyStandardLibraries/examples/BoundedIntExamples.dfy (hunk 2 above)
Comment (keyboardDrummer) [reply Source/DafnyStandardLibraries/examples/BoundedIntExamples.dfy:19]: My impression is that for Java an extern should not be in the default class since there are no partial classes.

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/wishlist/GoModule.dfy [hunk 4]
```diff
@@ -2,7 +2,7 @@
 // RUN: %exits-with 3 %dafny /compile:3 /unicodeChar:0 /spillTargetCode:2 "%s" /compileTarget:go 2> "%t"
 // note: putting /compileTarget:go after "%s" overrides user-provided option
 // RUN: %OutputCheck --file-to-check "%t" "%s"
-// CHECK: GoModuleConversions.go:10:3: "net/url" imported and not used
+// CHECK: undefined: GoModuleConversions.ParseURL
```
Comment (robin-aws) [inline Source/IntegrationTests/TestFiles/LitTests/LitTest/wishlist/GoModule.dfy:5]: Not clear on why this is the error we're getting now, but not a blocker since this is just a negative test.

File: Source/DafnyCore/Compilers/GoLang/Compiler-go.cs [hunk 5]
```diff
@@ -202,11 +197,32 @@ private struct Import {
...
-          importDummyWriter.WriteLine("var _ = os.Args");
+      bool isType;
+      string memberName = null;
+      if (id == "os") {
+        memberName = "Args";
+        isType = false;
+      } else {
+        isType = true;
+        if (import.ExternModule != null) {
+          var attributes = Attributes.Find(import.ExternModule.Attributes, "dummyImportMember");
```
Comment (robin-aws) [inline Source/DafnyCore/Compilers/GoLang/Compiler-go.cs:208]: Definitely glad we're not documenting this officially, but it might be worth adding a note on this attribute to https://github.com/dafny-lang/dafny/blob/master/Source/DafnyStandardLibraries/CONTRIBUTING.md, just so anyone that might need to use it is at least aware of how it works.

File: Source/DafnyCore/Compilers/SinglePassCompiler.cs [hunk 6]
```diff
@@ -167,12 +167,14 @@ public abstract class SinglePassCompiler {
     /// call to the instance Main method in the enclosing class.
     /// </summary>
     protected abstract ConcreteSyntaxTree CreateStaticMain(IClassWriter wr, string argsParameterName);
-    protected abstract ConcreteSyntaxTree CreateModule(string moduleName, bool isDefault, bool isExtern, string/*?*/ libraryName, ConcreteSyntaxTree wr);
+    protected abstract ConcreteSyntaxTree CreateModule(string moduleName, bool isDefault, ModuleDefinition externModule,
```
Comment (robin-aws) [inline Source/DafnyCore/Compilers/SinglePassCompiler.cs:170]: Definitely feels a little weird to have a parameter that is "the module being created, but only passed if it's an extern". Consider keeping `isExtern` and `theModule` as separate parameters.

File: Source/DafnyCore/AST/Members/ICodeContext.cs
Comment (robin-aws) [inline Source/DafnyCore/AST/Members/ICodeContext.cs:1]: Just for my own understanding, are the AST changes unrelated cleanup or necessary somehow?

**Review by robin-aws — APPROVED** [review]
Just a few more nits, nothing hard-blocking. Happy with the `{:dummyImportMember}` name. Do update the PR description so the commit message is accurate. :)

File: Source/DafnyCore/AST/Members/ICodeContext.cs
Comment (keyboardDrummer) [reply Source/DafnyCore/AST/Members/ICodeContext.cs:1]: Unrelated, yes
